File: src/main/java/com/aau/wizard/service/impl/GameManager.java
package com.aau.wizard.service.impl;

import com.aau.wizard.model.Card;
import com.aau.wizard.model.CardType;
import com.aau.wizard.model.Deck;
import com.aau.wizard.model.Suit;
import com.aau.wizard.model.PlayerState;
import com.aau.wizard.util.Pair;
import com.aau.wizard.util.TrickRules;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class GameManager {

    final List<PlayerState> players;
    final Deck deck = new Deck();
    Card trumpCard = null;
    Suit trumpSuit = null;
    final List<Pair<PlayerState, Card>> playedCards = new ArrayList<>();
    int currentTrickNumber = 0;

    public GameManager(List<PlayerState> players) {
        this.players = players;
    }

    public void startRound(int roundNumber) {
        deck.shuffle();
        for (PlayerState player : players) {
            player.setHand(new ArrayList<>(deck.draw(roundNumber)));
            player.setTricksWon(0);
            player.setBid(0);
        }

        if (deck.size() < 1) {
            trumpCard = null;
            trumpSuit = null;
        } else {
            List<Card> drawn = deck.draw(1);
            trumpCard = drawn.isEmpty() ? null : drawn.get(0);
            trumpSuit = trumpCard != null ? trumpCard.getSuit() : null;
        }

        currentTrickNumber = 0;
        playedCards.clear();

        System.out.println("Trumpf: " + (trumpSuit != null ? trumpSuit : "Kein Trumpf"));
    }

    public void playCard(PlayerState player, Card card) {
        if (!player.getHand().contains(card)) {
            throw new IllegalArgumentException("Player doesn't have that card");
        }

        if (!playedCards.isEmpty() && !TrickRules.isValidPlay(player, card, playedCards)) {
            throw new IllegalStateException("Invalid card play: " + card + " by " + player.getName());
        }

        player.getHand().remove(card);
        playedCards.add(new Pair<>(player, card));
    }

    public PlayerState endTrick() {
        if (playedCards.isEmpty()) {
            throw new IllegalStateException("No cards played in this trick");
        }

        PlayerState winner = TrickRules.determineTrickWinner(playedCards, trumpSuit);
        winner.setTricksWon(winner.getTricksWon() + 1);
        System.out.println("Stich " + (currentTrickNumber + 1) + " gewonnen von " + winner.getName() + " (" + winner.getTricksWon() + " Stiche)");

        playedCards.clear();
        currentTrickNumber++;
        return winner;
    }

    public void endRound() {
        for (PlayerState player : players) {
            int difference = Math.abs(player.getTricksWon() - player.getBid());
            int score;
            if (difference == 0) {
                score = 20 + (player.getTricksWon() * 10);
            } else {
                score = (player.getTricksWon() * 10) - (difference * 10);
            }
            player.setScore(player.getScore() + score);
        }

        System.out.println("\n=== Finale Auswertung ===");
        for (PlayerState player : players) {
            System.out.println(player.getName() + ": " + player.getBid() + " geboten, " +
                    player.getTricksWon() + " gewonnen → Punkte: " + player.getScore());
        }
    }
}

File: src/main/kotlin/com/aau/wizard/core/game/GameManager.kt
package com.aau.wizard.core.game

import com.aau.wizard.core.cards.Card
import com.aau.wizard.core.cards.CardType
import com.aau.wizard.core.cards.Deck
import com.aau.wizard.core.cards.Suit
import com.aau.wizard.core.model.PlayerState
import com.aau.wizard.core.rules.TrickRules
import kotlin.math.abs

class GameManager(internal val players: List<PlayerState>) {

    //Habe die variablen auf internal gemacht, um die Tests einfacher zu machen...
    internal val deck = Deck()
    internal var trumpCard: Card? = null
    var trumpSuit: Suit? = null
    internal val playedCards = mutableListOf<Pair<PlayerState, Card>>()
    internal var currentTrickNumber = 0


    fun startRound(roundNumber: Int) {
        deck.shuffle()
        players.forEach { player ->
            player.hand = deck.draw(roundNumber).toMutableList()
            player.tricksWon = 0
            player.bid = 0
        }

        //wenn letzte Runde passiert:
        if (deck.size() < 1){
            trumpCard = null
            trumpSuit = null
        }
        else{
            trumpCard = deck.draw(1).firstOrNull()
            trumpSuit = trumpCard?.suit
        }

        currentTrickNumber = 0
        playedCards.clear()

        println("Trumpf: ${trumpSuit ?: "Kein Trumpf"}")
    }


    fun playCard(player: PlayerState, card: Card) {
        require(player.hand.contains(card)) { "Player doesn't have that card" }

        if (playedCards.isNotEmpty() && !TrickRules.isValidPlay(player, card, playedCards)) {
            throw IllegalStateException("Invalid card play: $card by ${player.name}")
        }

        player.hand.remove(card)
        playedCards.add(player to card)
    }


    fun endTrick(): PlayerState {
        require(playedCards.isNotEmpty()) { "No cards played in this trick" }

        val winner = TrickRules.determineTrickWinner(playedCards, trumpSuit).also {
            it.tricksWon += 1
            println("Stich ${currentTrickNumber + 1} gewonnen von ${it.name} (${it.tricksWon} Stiche)")
        }

        playedCards.clear()
        currentTrickNumber++
        return winner
    }

    fun endRound() {
        players.forEach { player ->
            val difference = abs(player.tricksWon - player.bid)
            player.score += when {
                difference == 0 -> 20 + (player.tricksWon * 10)
                else -> (player.tricksWon * 10) - (difference * 10)
            }
        }

        println("\n=== Finale Auswertung ===")
        players.forEach {
            println("${it.name}: ${it.bid} geboten, ${it.tricksWon} gewonnen → Punkte: ${it.score}")
        }
    }
}